### PR TITLE
common: fix daemon abnormal exit at parsing invalid arguments

### PR DESCRIPTION
**Pull request by Yan-waller**
Fixes: [http://tracker.ceph.com/issues/21338](http://tracker.ceph.com/issues/21365)
Signed-off-by: Yan Jun <yan.jun8@zte.com.cn>

### CODE REVIEW

**Comment by Yan-waller** [comment]
restest this please

File: src/common/config.cc [hunk 1]
```diff
@@ -536,9 +536,11 @@ int md_config_t::parse_option(std::vector<const char*>& args,
     std::string as_option("--");
     as_option += "debug_";
     as_option += subsys.get_name(o);
-    if (ceph_argparse_witharg(args, i, &val,
+    if (ceph_argparse_witharg(args, i, &val, *oss,
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/common/config.cc]: @trociny and @jdurgin what do you think?

File: src/common/config.cc [hunk 2]
```diff
@@ -492,7 +492,7 @@ int md_config_t::parse_argv(std::vector<const char*>& args)
       set_val_or_die("client_mountpoint", val.c_str());
     }
     else {
-      parse_option(args, i, NULL);
+      parse_option(args, i, &cerr);
```
[the remark this answers is not in the document]
Comment (Yan-waller) [reply src/common/config.cc]: As the original logic, `cerr` is finally passed in to va_ceph_argparse_witharg(,,,cerr,), I'm not sure is this a minimal change that conform to original intent ?..
```c++
parse_option(args, i, NULL); ->
ceph_argparse_witharg(args, i, &val,
                      as_option.c_str(), (char*)NULL)); ->
va_ceph_argparse_witharg(args, i, ret, cerr, ap);
```

File: src/common/config.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (trociny) [reply src/common/config.cc]: Yes, `ceph_argparse_witharg()/ceph_argparse_binary_flag()` interface looks unfortunate. Still I am not sure it worth changing at this stage.

Although we can't use the return value to tell if parsing failed, it looks it is safe to assume it failed if something was printed to stream provides as `std::ostream &oss` param.

I.e. I think the reported problem could be fixed by something like this:
```
+    ostringstream err; 
+    if (ceph_argparse_witharg(args, i, &val, err,
                              as_option.c_str(), (char*)NULL)) {
+      if (!err.str().empty()) {
+        if (oss) {
+          *oss << err.str(); 
+        }
+        ret = -EINVAL;
+        break;
+      }
```

**Comment by Yan-waller** [comment]
@tchaikov I have fixed it as you and trociny's suggestions, could you help to take a look again? thanks.

File: src/common/config.cc [hunk 3]
```diff
@@ -536,8 +539,16 @@ int md_config_t::parse_option(std::vector<const char*>& args,
     std::string as_option("--");
     as_option += "debug_";
     as_option += subsys.get_name(o);
-    if (ceph_argparse_witharg(args, i, &val,
+    ostringstream err;
+    if (ceph_argparse_witharg(args, i, &val, err,
 			      as_option.c_str(), (char*)NULL)) {
+      if (!err.str().empty()) {
```
Comment (tchaikov) [inline src/common/config.cc]: nit, use `tellp()` instead of `err.str().empty()`

File: src/test/daemon_config.cc [hunk 4]
```diff
@@ -215,6 +215,11 @@ TEST(DaemonConfig, InjectArgsReject) {
   ret = g_ceph_context->_conf->get_val("osd_data", &tmp2, sizeof(buf2));
   ASSERT_EQ(ret, 0);
   ASSERT_EQ(string(buf), string(buf2));
+
+  // We should complain about the missing arguments.
+  std::string injection3("--log-graylog-port 28 --debug_ms");
+  ret = g_ceph_context->_conf->injectargs(injection3, &cout);
+  ASSERT_EQ(ret, -EINVAL);
```
Comment (tchaikov) [inline src/test/daemon_config.cc]:
```c
ASSERT_EQ(-EINVAL, ret);
```

as the 1st param is the expected value, the second one is the actual value.

**Comment by Yan-waller** [comment]
@tchaikov thanks, have done it. and incidentally, I adjust all ASSERT_EQ in `daemon_config.cc ` file as another commit.

**Comment by Yan-waller** [comment]
retest this please

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2017-09-14_12:06:50-rados-wip-kefu-testing-2017-09-14-1846-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2017-09-15_05:21:49-rados-wip-kefu-testing-2017-09-14-1846-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2017-09-15_10:57:17-rados-wip-kefu-testing-2017-09-14-1846-distro-basic-mira/

**Pull request merged**